---
title: "Bulk RNA-seq analysis (iTBS vs. cTBS vs. sham)"
author: "Rebecca Ong"
date: "02/11/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicFeatures)
library(rtracklayer)
library(S4Vectors)
library(Rsamtools)
library(GenomicAlignments)
library(BiocParallel)
library(pheatmap)
library(RColorBrewer)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(genefilter)
library(ggplot2)
library(ggrepel)
library(clusterProfiler)
library(DOSE)
library(dplyr)
library(tidyr)
library(Rgraphviz)
library(enrichplot)
library(stringr)
library(DESeq2)
library(ggupset)
```

# Data Import
A '.csv' sample info table was created and loaded into R. STAR aligned files were then locaated and BAM files to be used for counting was specified.
```{r}
csvfile <- file.path("/bulkrtms/Raw_files_AGRF_CAGRF220711399_HVKMLDSX3", "Sample_table.csv")
sampleTable <- read.csv(csvfile,row.names = 1)
sampleTable
sampleTable$IDntreatment <- c('B001_iTBS', 'B002_cTBS','B003_iTBS', 'B004_sham', 'B005_cTBS', 'B006_sham', 'B007_sham', 'B008_cTBS', 'B009_iTBS', 'B010_iTBS', 'B011_iTBS', 'B012_cTBS', 'B013_cTBS', 'B014_sham', 'B015_sham')
setwd("/bulkrtms/Merged_files/alignment_adjusted")
aligned_dir <- file.path("/bulkrtms/Merged_files/alignment_adjusted")
list.files(aligned_dir)
filenames <- file.path(aligned_dir, paste0(sampleTable$SampleID, "Aligned.sortedByCoord.out.bam"))
file.exists(filenames)
bamfile <- BamFileList(filenames)
seqinfo(bamfile[[1]])
```

## Reference genome annotation
Import the mouse reference genome for annotation
```{r}
GTFfile <- file.path("/bulkrtms/referencegenome", "Mus_musculus.GRCm39.107.gtf")
txdb <- makeTxDbFromGFF(GTFfile, format="gtf", circ_seqs = character())
exonsbygene <- exonsBy(txdb, by = 'gene')
exonsbygene
```

## Read counting
Read counting step. In this step, we generate a 'SummarizedExperiment' object that sumarrises the RNA-seq experiment. 
  - Each sample is in the column -> colData(se)
  - Each gene is contain in the rows -> rowData(se) lists the p values and log changes of each ene after differential expression analysis
```{r}
register(SerialParam())
se <- summarizeOverlaps(features=exonsbygene, reads = bamfile, mode = "Union", singleEnd=FALSE, ignore.strand = TRUE, fragment=FALSE)
colnames(se) <- sampleTable$IDntreatment
se
dim(se)
assayNames(se)
head(assay(se),3)
colSums(assay(se))
rowRanges(se)
colData(se)
se$SampleID
se$Treatment
```

## The DESeqDataSet
From the 'SummarizedExperiment', we can create a 'DESeqDataSet' to store the read counts and intermediate estimated quantities during the statistical analysis from the DESeq2 package.
```{r}
dds <- DESeqDataSet(se, design = ~ Treatment)
dds
```

Before we perform differential expression analysis, we removed rows that had no reads detected (i.e., we keep only rows that have more than 0 reads).
```{r}
dds <- dds[ rowSums(counts(dds)) > 0, ]
dds
```

Next we define the reference level that we want to compare our stimulation datasets against (i.e., the level that represents the control group).
```{r}
dds$Treatment <- relevel(dds$Treatment, ref = "sham")
nrow(dds)
```

# Data Quality Assessment
Because the purpose of this analysis is to detect deferentially expressed genes, we need to have a look at the quality of the data set for particular samples that may have abnormalities due to slight differences in the experimental treatment/protocol.

## Bar plot of library sizes
Check how many reads we have for each sample.
```{r}
librarySizes <- colSums(assay(dds))
dds$IDntreatment
barplot(librarySizes/1e06, names=dds$SampleID, las = 2, cex.names = 0.8, xlab = "Samples", ylab = "Library size (millions)", main = "Barplot of library sizes")
```
## Log2 of counts
Count data is not normally distributed so to estimate the distribution of raw counts, we need to log the counts.
```{r}
logcounts <- log2(assay(dds) + 1)
boxplot(logcounts, xlab ="", ylab = "Log2 counts per million", las=2)
abline(h=median(as.matrix(logcounts)), col = "blue")
```
## VST transformed counts box plot
```{r}
vsd <- vst(dds, blind = FALSE)
head(assay(vsd),3)
colData((vsd))
boxplot(assay(vsd), xlab = "", ylab = "vst tranformed counts", las = 2)
```

# Differential Expression
## PCA Plot
Differential expression is performed on RAW counts but for downstream analysis (i.e., visualisation or clustering) it is useful to work with transformed versions of the count data (e.g., variance stabilising transformation (VST)).
A VST transformation takes into account library size and the point is to remove the dependence of the variance on the mean.
```{r}
plotPCA(vsd, intgroup = "Treatment")
```
Customise the PCA plots using ggplot functions
```{r}
PCAplot <- plotPCA(vsd, intgroup = "Treatment", returnData = TRUE)
percentVar <- round(100 * attr(PCAplot, "percentVar"))
ggplot(PCAplot, aes(x = PC1, y = PC2, color = Treatment)) + geom_point(size = 4) + xlab(paste0("PC1: ", percentVar[1], "% variance")) + ylab(paste0("PC2: ", percentVar[2], "% variance")) + coord_fixed() + scale_color_manual(values = c("grey50", "dodgerblue1", "palevioletred1"))
```

## Differential expression analysis
The 'DESeq' function includes all the standard differential expression analysis steps. The results table can be extracted with log2 fold changes, p values and adjusted p values.
```{r}
dds$Treatment <- relevel(dds$Treatment, ref='sham')
dds <- DESeq(dds)
```
### DEG Results (iTBS vs sham)
Summary of the results of differential gene expression analysis between iTBS and sham.
```{r}
resultsNames(dds)
res_iTBS <- results(dds, contrast = list("Treatment_iTBS_vs_sham"))
res_iTBS
res_iTBS <- data.frame(res_iTBS)
summary(res_iTBS)
```

### DEG Results (cTBS vs sham)
Summary of the results of differential gene expression analysis between cTBS and sham.
```{r}
res_cTBS <- results(dds, contrast = list("Treatment_cTBS_vs_sham"))
res_cTBS
summary(res_cTBS)
```

### Annotating & exporting results
```{r}
columns(org.Mm.eg.db)
res_cTBS$symbol <- mapIds(org.Mm.eg.db, keys=row.names(res_cTBS), column = "SYMBOL", keytype = "ENSEMBL", multiVals = "first")
res_cTBS$entrez <- mapIds(org.Mm.eg.db, keys=row.names(res_cTBS), column = "ENTREZID", keytype = "ENSEMBL", multiVals = "first")
res_cTBS$genename <- mapIds(org.Mm.eg.db, keys=row.names(res_cTBS), column = "GENENAME", keytype = "ENSEMBL", multiVals = "first")

res_iTBS$symbol <- mapIds(org.Mm.eg.db, keys=row.names(res_iTBS), column = "SYMBOL", keytype = "ENSEMBL", multiVals = "first")
res_iTBS$entrez <- mapIds(org.Mm.eg.db, keys=row.names(res_iTBS), column = "ENTREZID", keytype = "ENSEMBL", multiVals = "first")
res_iTBS$genename <- mapIds(org.Mm.eg.db, keys=row.names(res_iTBS), column = "GENENAME", keytype = "ENSEMBL", multiVals = "first")
```

Re-order results table by the smallest p value
```{r}
res_cTBSOrdered <- res_cTBS[order(res_cTBS$pvalue),]
head(res_cTBSOrdered)
summary(res_cTBSOrdered)

res_iTBSOrdered <- res_iTBS[order(res_iTBS$pvalue),]
head(res_iTBSOrdered)
summary(res_iTBSOrdered)
```

Remove all the genes with no entrez ID or name (i.e., NAs)
```{r}
res_cTBSOrdered_df <- data.frame(res_cTBSOrdered)
res_cTBSOrdered_noNA <- res_cTBSOrdered_df %>% drop_na()
summary(res_cTBSOrdered_noNA)

res_iTBSOrdered_df <- data.frame(res_iTBSOrdered)
res_iTBSOrdered_noNA <- res_iTBSOrdered_df %>% drop_na()
```

### Significant DEGs (cTBS vs sham)
Set the significance cutoff values of padj < 0.05 and log2fc > 1.
```{r}
res_cTBSSig <- subset(res_cTBSOrdered, padj <= 0.05 & abs(log2FoldChange) >= 1)
res_cTBSSig
summary(res_cTBSSig)

# remove all Sig genes with NA
res_cTBSSig_noNA <- subset(res_cTBSOrdered_noNA, padj < 0.05 & abs(log2FoldChange) > 1)
summary(res_cTBSSig_noNA)
```

### Significant DEGs (iTBS vs sham)
Set the significance cutoff values of padj < 0.05 and log2fc > 1.
```{r}
res_iTBSSig <- subset(res_iTBSOrdered, padj <= 0.05 & abs(log2FoldChange) >= 0.5)
res_iTBSSig
summary(res_iTBSSig)
```

# Presenting DEG results
## Volcano plot (cTBS)
Create a volcano plot labeling the top significant genes following cTBS stimulation.
```{r}
res_cTBSOrdered_df <- res_cTBSOrdered_df %>% 
  mutate(
    Expression = case_when(log2FoldChange >= 1 & padj <= 0.05 ~ "Up-regulated",
                           log2FoldChange <= -1 & padj <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged")
  )
# without labels
cTBS_volcano <- ggplot(res_cTBSOrdered_df, aes(log2FoldChange, -log(padj,10))) + geom_point(aes(color = Expression), size = 2.5) + xlab(expression("log"[1]*"FC")) + ylab(expression("-log"[10]*"padj")) + scale_color_manual(values = c("tomato2", "gray50", "yellowgreen")) + guides(colour = guide_legend(override.aes = list(size=1.5))) + xlim(-5,5)
cTBS_volcano

# add in dotted lines to indicate where the significance cutoffs were
cTBS_volcano_lines <- cTBS_volcano + geom_vline(xintercept = c(-1, 1), color = 'black', linetype = 2, size = 0.5) +  geom_hline(yintercept = -log10(0.05), color = 'black', linetype = 2, size = 0.5)
cTBS_volcano_lines
cTBS_volcano_lines <- cTBS_volcano + geom_hline(yintercept = -log10(0.05), color = 'black', linetype = 2, size = 0.5)
cTBS_volcano_lines
```

## Volcano plot (iTBS)
Create a volcano plot labeling the top significant genes following iTBS stimulation.
```{r}
res_iTBSOrdered_df <- res_iTBSOrdered_df %>% 
  mutate(
    Expression = case_when(log2FoldChange >= 1 & padj <= 0.05 ~ "Up-regulated",
                           log2FoldChange <= -1 & padj <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged")
  )
# without labels
iTBS_volcano <- ggplot(res_iTBSOrdered_df, aes(log2FoldChange, -log(padj,10))) + geom_point(aes(color = Expression), size = 2.5) + xlab(expression("log"[1]*"FC")) + ylab(expression("-log"[10]*"padj")) + scale_color_manual(values = c("gray50", "gray50", "gray50")) + guides(colour = guide_legend(override.aes = list(size=1.5))) + xlim(-5,5) + ylim(0,4)
iTBS_volcano

iTBS_volcano_lines <- iTBS_volcano + geom_vline(xintercept = c(-1, 1), color = 'black', linetype = 2, size = 0.5) +  geom_hline(yintercept = -log10(0.05), color = 'black', linetype = 2, size = 0.5)
iTBS_volcano_lines
```

# GO analysis
Identify the pathways that are differentially expressed following cTBS. The file 'res_cTBSSig_noNA' contains a list of the significant differentially expressed genes where all the NA values have been removed.
```{r}
cTBS_geneList <- res_cTBSOrdered_noNA[,2]
names(cTBS_geneList) <- as.character(res_cTBSOrdered_noNA[,8])
cTBS_geneList <- sort(cTBS_geneList, decreasing = TRUE)
de <- names(cTBS_geneList)
```
## Perform the GO enrichment analysis
```{r}
cTBS_CC <- enrichGO(de, OrgDb = org.Mm.eg.db, keyType = "ENTREZID", ont = "BP", pAdjustMethod = "BH", pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)
write.csv(as.data.frame(cTBS_CC),"GO enrich CC.csv",row.names =F)
```
NOTE: no GO pathways identified when performing GO analysis on SIGNIFICANT differentially expressed genes following cTBS
